renderer_cocoa/lib/Cocoa.ml

```diff
@@ -0,0 +1,228 @@
+      end
+    end
+
+  type delegateSelector =
+    | ApplicationWillFinishLaunching
+    | ApplicationDidFinishLaunching
+
+  let applicationDelegate id sel =
+    try
+      let del = T.find application_table id in
+      match sel with
+      | ApplicationWillFinishLaunching ->
+          del.applicationWillFinishLaunching ()
+      | ApplicationDidFinishLaunching ->
+          del.applicationDidFinishLaunching ()
+    with Not_found ->
+      ()
+
+  let () =
+    Callback.register "NSApp.delegate" applicationDelegate
+end
+
+module NSWindow = struct
+  external _NSWindow_makeWithContentRect: int -> float -> float -> float -> float -> nsWindow = "ml_NSWindow_makeWithContentRect"
+
+  external _NSWindow_isVisible: nsWindow -> bool = "ml_NSWindow_isVisible"
+  external _NSWindow_center: nsWindow -> unit = "ml_NSWindow_center"
+  external _NSWindow_makeKeyAndOrderFront:
+    nsWindow -> unit = "ml_NSWindow_makeKeyAndOrderFront"
+  external _NSWindow_setTitle: nsWindow -> string -> unit = "ml_NSWindow_setTitle"
+  external _NSWindow_title: nsWindow -> string = "ml_NSWindow_title"
+  external _NSWindow_contentView: nsWindow -> nsView = "ml_NSWindow_contentView"
+  external _NSWindow_setContentView: nsWindow -> nsView -> unit = "ml_NSWindow_setContentView"
+
+  class type t =
+    object
+      method isVisible: bool
+      method center: unit
+      method makeKeyAndOrderFront: unit
+      method setTitle: string -> unit
+      method title: string
+
+      method contentView: nsView
+      method setContentView: nsView -> unit
+
+      method windowDidResize: (unit -> unit) -> unit
+    end
+
+  type windowDelegate = {
+    mutable windowDidResize: (unit -> unit);
+  }
+
+  let delegate () = {
+    windowDidResize = (fun () -> ());
+  }
+
+  let win_id = ref 0
+  let window_table = T.create 0
+
+  let makeWithContentRect x y w h =
+    let id = incr win_id; !win_id in
+    let win = _NSWindow_makeWithContentRect id x y w h in
+    let del = delegate () in
+    T.add window_table id del;
+    object
+      method isVisible = _NSWindow_isVisible win
+      method center = _NSWindow_center win
+      method makeKeyAndOrderFront = _NSWindow_makeKeyAndOrderFront win
+      method setTitle s = _NSWindow_setTitle win s
+      method title = _NSWindow_title win
+
+      method contentView = _NSWindow_contentView win
+      method setContentView v = _NSWindow_setContentView win v
+
+      method windowDidResize f =
+        del.windowDidResize <- f
+    end
+
+  type windowDelegateSelector =
+    | WindowDidResize
+
+  let windowDelegate id sel =
+    try
+      let del = T.find window_table id in
+      match sel with
+      | WindowDidResize ->
+          del.windowDidResize ()
+    with Not_found ->
+      Printf.ksprintf prerr_endline "Window #%d has been GCed" id
+
+  let () =
+    Callback.register "NSWindow.delegate" windowDelegate
+end
+
+module NSView = struct
+  type t = nsView
+
+  external _NSView_make: unit -> t = "ml_NSView_make"
```

Comment:
That said it's interesting how they bind to `NSView`? How did they make sure it says in the memory after passing the pointer to OCaml.

Reply:
I believe OCaml is just wrapping `NSView` pointer.

https://github.com/briskml/brisk/blob/ac4b5d449ea2dcc09a3ab08f2dcf149d2a374cbd/renderer_cocoa/lib/Cocoa_stubs.c#L58-L59

Reply:
Yeah, I checked the stubs. Could you point me to the original source?

Reply:
There's nothing on NSView though:
https://github.com/nojb/ocaml-cocoa/blob/master/lib/cocoa_stubs.m

Reply:
Ok, I'm generally for simplifying those bindings as much as possible.

Reply:
I'll take a sweep in a separate PR and noalloc-ize the bindings where possible.